assembly/src/tokens/stream.rs

```diff
@@ -26,18 +27,23 @@ impl<'a> TokenStream<'a> {
         }
         let mut tokens = Vec::new();
         let mut doc_comments = BTreeMap::new();
+        let mut module_comment = None;
 
-        let mut comment = Comment(None);
+        let mut doc_comment = Comment(None);
 
         for line in source.lines() {
             let line = line.trim();
             if line.starts_with(DOC_COMMENT_PREFIX) {
-                comment.append_line(line);
-            } else if line.is_empty() && !comment.is_empty() {
-                return Err(ParsingError::malformed_doc_comment(tokens.len()));
+                doc_comment.append_line(line);
+            } else if line.is_empty() && !doc_comment.is_empty() {
+                if tokens.is_empty() && module_comment.is_none() {
```

Comment:
nit: I would probably add a small comment here to explain this. Something like: "If we haven't read any tokens yet, but already have built a comment, a new line must indicate the end of a module comment."